### PR TITLE
librados: make querying pools for selfmanaged snaps reliable

### PR DESCRIPTION
**Pull request by idryomov**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by idryomov** [comment]
`qa/workunits/mon/rbd_snaps_ops.sh` workunit is run from the RBD suite.

Updated workunit without the fix:

```
2023-11-23T20:44:07.310 INFO:tasks.workunit.client.0.smithi022.stdout:[23] rados cppool test-foo test-bar
2023-11-23T20:44:07.346 INFO:tasks.workunit.client.0.smithi022.stderr:WARNING: pool copy does not preserve user_version, which some apps may rely on.
2023-11-23T20:44:07.355 INFO:tasks.workunit.client.0.smithi022.stdout:test-foo:rbd_id.image => test-bar:rbd_id.image
2023-11-23T20:44:07.357 INFO:tasks.workunit.client.0.smithi022.stdout:test-foo:rbd_header.109513078af9 => test-bar:rbd_header.109513078af9
2023-11-23T20:44:07.359 INFO:tasks.workunit.client.0.smithi022.stdout:test-foo:rbd_directory => test-bar:rbd_directory
2023-11-23T20:44:07.361 INFO:tasks.workunit.client.0.smithi022.stdout:test-foo:rbd_object_map.109513078af9 => test-bar:rbd_object_map.109513078af9
2023-11-23T20:44:07.363 INFO:tasks.workunit.client.0.smithi022.stdout:test-foo:rbd_object_map.109513078af9.0000000000000002 => test-bar:rbd_object_map.109513078af9.0000000000000002
2023-11-23T20:44:07.366 INFO:tasks.workunit.client.0.smithi022.stdout:successfully copied pool test-foo
2023-11-23T20:44:07.369 DEBUG:teuthology.orchestra.run:got remote process result: 1
2023-11-23T20:44:07.370 INFO:tasks.workunit.client.0.smithi022.stdout:[23] unexpected return '0', expected '1'
```

https://pulpito.ceph.com/dis-2023-11-23_20:19:59-rbd-main-distro-default-smithi/

Updated workunit with the fix:

https://pulpito.ceph.com/dis-2023-11-24_04:23:13-rbd-wip-dis-testing-distro-default-smithi/

**Comment by idryomov** [comment]
With the fixup, segfault is gone (ran 10 jobs to make sure):

```
2023-11-24T22:53:05.037 INFO:tasks.workunit.client.0.smithi204.stdout:[23] rados cppool test-foo test-bar
2023-11-24T22:53:05.072 INFO:tasks.workunit.client.0.smithi204.stderr:WARNING: pool copy does not preserve user_version, which some apps may rely on.
2023-11-24T22:53:05.073 INFO:tasks.workunit.client.0.smithi204.stderr:WARNING: pool test-foo has selfmanaged snaps, which are not preserved
2023-11-24T22:53:05.073 INFO:tasks.workunit.client.0.smithi204.stderr:    by the cppool operation.  This will break any snapshot user.
2023-11-24T22:53:05.073 INFO:tasks.workunit.client.0.smithi204.stderr:    If you insist on making a broken copy, you can pass
2023-11-24T22:53:05.073 INFO:tasks.workunit.client.0.smithi204.stderr:    --yes-i-really-mean-it to proceed anyway.
2023-11-24T22:53:05.076 INFO:tasks.workunit.client.0.smithi204.stdout:[24] rados cppool test-foo test-bar --yes-i-really-mean-it
2023-11-24T22:53:05.107 INFO:tasks.workunit.client.0.smithi204.stderr:WARNING: pool copy does not preserve user_version, which some apps may rely on.
2023-11-24T22:53:05.108 INFO:tasks.workunit.client.0.smithi204.stderr:WARNING: pool test-foo has selfmanaged snaps, which are not preserved
2023-11-24T22:53:05.108 INFO:tasks.workunit.client.0.smithi204.stderr:    by the cppool operation.  This will break any snapshot user.
2023-11-24T22:53:05.114 INFO:tasks.workunit.client.0.smithi204.stdout:test-foo:rbd_id.image => test-bar:rbd_id.image
2023-11-24T22:53:05.117 INFO:tasks.workunit.client.0.smithi204.stdout:test-foo:rbd_header.1094b14e6b55 => test-bar:rbd_header.1094b14e6b55
2023-11-24T22:53:05.119 INFO:tasks.workunit.client.0.smithi204.stdout:test-foo:rbd_directory => test-bar:rbd_directory
2023-11-24T22:53:05.121 INFO:tasks.workunit.client.0.smithi204.stdout:test-foo:rbd_object_map.1094b14e6b55.0000000000000002 => test-bar:rbd_object_map.1094b14e6b55.0000000000000002
2023-11-24T22:53:05.123 INFO:tasks.workunit.client.0.smithi204.stdout:test-foo:rbd_object_map.1094b14e6b55 => test-bar:rbd_object_map.1094b14e6b55
2023-11-24T22:53:05.125 INFO:tasks.workunit.client.0.smithi204.stdout:successfully copied pool test-foo
```
https://pulpito.ceph.com/dis-2023-11-24_22:29:39-rbd-wip-dis-testing-distro-default-smithi/

**Review by ajarr — APPROVED** [review]

File: qa/workunits/mon/rbd_snaps_ops.sh [hunk 1]
```diff
@@ -52,6 +53,8 @@ expect 'rbd --pool test-foo snap create image@snapshot' 0
 ceph osd pool delete test-bar test-bar --yes-i-really-really-mean-it || true
 expect 'ceph osd pool create test-bar 8' 0
 expect 'ceph osd pool application enable test-bar rbd'
+# "rados cppool" without --yes-i-really-mean-it should fail
+expect 'rados cppool test-foo test-bar' 1
```
Comment (rzarzynski) [inline qa/workunits/mon/rbd_snaps_ops.sh:57]: ACK.

File: src/tools/rados/rados.cc [hunk 2]
```diff
@@ -3120,15 +3120,21 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
...
 	 << "apps may rely on." << std::endl;
 
-    if (rados.get_pool_is_selfmanaged_snaps_mode(src_pool)) {
+    ret = rados.pool_is_in_selfmanaged_snaps_mode(src_pool);
+    if (ret < 0) {
+      cerr << "failed to query pool " << src_pool << " for selfmanaged snaps: "
+           << cpp_strerror(ret) << std::endl;
+      return 1;
+    }
+    if (ret > 0) {
```
Comment (rzarzynski) [inline src/tools/rados/rados.cc:3129]: nit: after seeing the initial `if(ret < 0)` I started doubke0-checking whether the block above doesn't modify `ret`. How about using `else` instead or making `ret` a constant?

Comment (idryomov) [reply src/tools/rados/rados.cc:3129]: > How about using `else` instead

That would be an "else after return", which is thrown upon by many.

> or making `ret` a constant?

`ret` is declared ~1200 lines above ;)

Comment (rzarzynski) [reply src/tools/rados/rados.cc:3129]: > That would be an "else after return", which is thrown upon by many.

Yes but right now we have:

```cpp
if (ret < 0) {
  return;
}
if (ret > 0) {
  // ...
}
```

which, from the point of indentation, has all the disadvantages of `else`  but additionally it draws extra attention. How about getting rid of it?

Comment (idryomov) [reply src/tools/rados/rados.cc:3129]: > How about getting rid of it?

You said "yes" to the style concern, so I'm not sure what you mean by "get rid of it".  Is
```
if (ret < 0) {
  // ...
  return;
} else if (ret > 0) {
  // ...
}
```
what you would rather see?

File: src/include/rados/librados.hpp [hunk 3]
```diff
@@ -1477,8 +1477,11 @@ inline namespace v14_2_0 {
     int get_pool_stats(std::list<std::string>& v,
                        std::string& category,
 		       std::map<std::string, stats_map>& stats);
-    /// check if pool has selfmanaged snaps
-    bool get_pool_is_selfmanaged_snaps_mode(const std::string& poolname);
+
+    /// check if pool has or had selfmanaged snaps
+    bool get_pool_is_selfmanaged_snaps_mode(const std::string& poolname)
+      __attribute__ ((deprecated));
+    int pool_is_in_selfmanaged_snaps_mode(const std::string& poolname);
```
Comment (gregsfortytwo) [inline src/include/rados/librados.hpp:1484]: Unless I'm misunderstanding, this is removing an existing function without a deprecation period.

Do we version librados? (I think we do?) I know we're not good about revving it, but this should increment the version.

I realize the old version is unsafe, but we should at least make it a pass-thru to the new safe implementation, perhaps tweaking the return value on pool ENOENT cases.

**Review by gregsfortytwo — CHANGES_REQUESTED** [review]

File: src/tools/rados/rados.cc (hunk 2 above)
Comment (rzarzynski) [reply src/tools/rados/rados.cc:3129]: Yes but it's still a nit, so I'm not going to fight for it :-). The intention behind is to somehow tell a reader these are 2 checks for the same call / assignment. IMHO 2 separated `if` suggest an extra assignment happens in the first check.

File: src/include/rados/librados.hpp (hunk 3 above)
Comment (idryomov) [reply src/include/rados/librados.hpp:1484]: > we should at least make it a pass-thru to the new safe implementation, perhaps tweaking the return value on pool ENOENT cases.

This is exactly what is done here:
```
// deprecated, use pool_is_in_selfmanaged_snaps_mode() instead
bool librados::Rados::get_pool_is_selfmanaged_snaps_mode(const std::string& pool)
{
  return client->get_pool_is_selfmanaged_snaps_mode(pool);
  // errors are ignored, prone to false negative results
  return client->pool_is_in_selfmanaged_snaps_mode(pool) > 0;
}
```

Comment (gregsfortytwo) [reply src/include/rados/librados.hpp:1484]: *sigh* sorry for the noise, I saw this on my phone and missed the librados_cxx change and got confused by the RadosClient one.

File: src/tools/rados/rados.cc (hunk 2 above)
Comment (idryomov) [reply src/tools/rados/rados.cc]: Obliged ;)

**Comment by idryomov** [comment]
jenkins retest this please

**Comment by ronen-fr** [comment]
RADOS approves:
https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomc6sy6WQmn1915-wip-yuri10-testing-2023-12-23-0824

**Pull request merged**